packages/react-dom/src/client/ReactDOMHostConfig.js

```diff
@@ -1117,3 +1127,142 @@ export function preparePortalMount(portalInstance: Instance): void {
     listenToEvent('onMouseEnter', portalInstance);
   }
 }
+
+export const supportsTestSelectors = true;
+
+export function findFiberRoot(node: Instance): null | FiberRoot {
+  const stack = [node];
+  let index = 0;
+  while (index < stack.length) {
+    const current = stack[index++];
+    if (isContainerMarkedAsRoot(current)) {
+      return ((getInstanceFromNodeDOMTree(current): any): FiberRoot);
+    }
+    stack.push(...current.children);
+  }
+  return null;
+}
+
+export function getBoundingRect(node: Instance): BoundingRect {
+  const rect = node.getBoundingClientRect();
+  return {
+    x: rect.left,
+    y: rect.top,
+    width: rect.width,
+    height: rect.height,
+  };
+}
+
+export function matchAccessibilityRole(node: Instance, role: string): boolean {
+  if (role === getRole(node)) {
+    return true;
+  }
+
+  return false;
+}
+
+export function getTextContent(fiber: Fiber): string | null {
+  switch (fiber.tag) {
+    case HostComponent:
+      let textContent = '';
+      const childNodes = fiber.stateNode.childNodes;
```

Comment:
Why can't we do `fiber.stateNode.textContent` here? What does it miss compared to this approach?

Reply:
Great question!

I think that would make the text selector type useless, since it would always match every root (since `textContent` includes all of the DOM subtree). The way this is implemented, we will only return the contents of the *immediate* `TEXT_NODE` children. (There's a test that verifies this behavior.) I think this makes the selector actually useful.